File: p12/p12.py
#!/usr/bin/env python

## What is the value of the first triangle number to have over five hundred divisors?
## http://projecteuler.net/index.php?section=problems&id=12

import sys
from math import sqrt

if len(sys.argv) < 3:
	print "usage: ", sys.argv[0], " start end [quiet]"
	sys.exit(1)

updates = True
if len(sys.argv) > 3:
	updates = False

# t(tnum, i) - returns the next triangular number #
def t(tnum, i):
	i = i + 1
	tnum = tnum + i
	return tnum, i

# factorise(n) - returns a list of factors of n #
def factorise(n, prev_nums):
	if n < 2:
		return [1]
	lim = int(sqrt(n))
	factors = dict([(i, False) for i in range(2, lim + 1)]) # nothing is a factor yet
	factors[1] = True # 1 is a factor of everything, bar 0
	factors[n] = True
	f_keys = factors.keys()
	f_keys.sort(reverse=True)
	for i in f_keys:
		if factors[i]:
			continue
		elif (n % i == 0):
			if i in prev_nums.keys():
				for j in prev_nums[i]:
					factors[j] = True
			else:
				factors[i] = (n % i == 0)
	fac_list = []
	for i in factors.keys():
		if factors[i]:
			fac_list.append(i)
	return fac_list

n = int(sys.argv[1])
end = int(sys.argv[2])

print "Searching for a triangular number with over 500 factors"
print "Checking triangle numbers from T[", n, "] - T[", (end - 1), "]"

factors = {}
max = 0
index = n
curr_t = n * (n + 1) / 2

for i in range(n, end):
	factors[curr_t] = factorise(curr_t, factors)
	
	if len(factors[curr_t]) > max:
		max = len(factors[curr_t])
		print "T[", i, "]:", curr_t, " - max found: ", max
	
	if updates:
		print "T[", i, "]: ", curr_t, " has ", len(factors[curr_t]), " factors"
		print factors[curr_t] # this generates a LOT of output!
	
	if len(factors[curr_t]) > 500:
		print "Finished! ", curr_t, " has ", len(factors[curr_t]), " divisors."
		break
	curr_t,index = t(curr_t, index)

File: p12/p12.pl
#!/usr/bin/perl

###
### NOTE: this solution is inefficient!
### See p12.py for a version that uses a lookup
### table of previous calculated factors...
###


## What is the value of the first triangle number to have over five hundred divisors?
## http://projecteuler.net/index.php?section=problems&id=12

use strict;
use warnings;

# subs #
sub trianglenumbergenerator
{
	my $index = 0;
	my $trianglenumber = 0;
	return sub { ++$index; $trianglenumber += $index; return $trianglenumber; };
}

sub divisors
{
	my ($num) = @_;
	my $divisors = 1; # include the number itself
	for my $i(1 .. ($num / 2))
	{
		if($num % $i == 0)
		{
			++$divisors;
		}
	}

	return $divisors;
}

# main #
print "finding first triangular number with over five hundred divisors..\n";

my $hardlimit = 10000; # failsafe ;)
my ($numdivisors, $start) = 500; # the number of divisors we want
my $nexttriangle = trianglenumbergenerator();
my $highest_t = 0;
my $highest_d = 0;
my $last_output = 0;

# find the solution! #
for my $i($start .. $hardlimit)
{
	my $trianglenumber = $nexttriangle->();
	my $divisors = divisors($trianglenumber);

	if($divisors > $highest_d)
	{
		$last_output = 0;
		$highest_d = $divisors;
		$highest_t = $trianglenumber;
		print 'highest num divisors is ' . $highest_d . ' at ' . $highest_t . '(' . ($highest_d * 100 / $numdivisors) . "% there)\n";
	}

	if($divisors > $numdivisors)
	{
		print "Found it!\nAnswer: " . $trianglenumber . "\n";
		last;
	}

	if($last_output > 1000)
	{
		$last_output = 0;
		print "Don't worry, folks - we're still searching! [" . $i . " / " . $hardlimit . "]\n";
	}

	$last_output++;
}
